contracts/PartyBid.sol

```diff
@@ -0,0 +1,701 @@
+        totalContributed[_contributor] =
+            totalContributed[_contributor] +
+            _amount;
+        // add to party's total contribution & emit event
+        totalContributedToParty = totalContributedToParty + _amount;
+        emit Contributed(
+            _contributor,
+            _amount,
+            _previousTotalContributedToParty,
+            totalContributed[_contributor]
+        );
+    }
+
+    // ======== External: Bid =========
+
+    /**
+     * @notice Submit a bid to the Market
+     * @dev Reverts if insufficient funds to place the bid and pay PartyDAO fees,
+     * or if any external auction checks fail (including if PartyBid is current high bidder)
+     * Emits a Bid event upon success.
+     * Callable by any contributor
+     */
+    function bid() external nonReentrant {
+        require(
+            partyStatus == PartyStatus.AUCTION_ACTIVE,
+            "PartyBid::bid: auction not active"
+        );
+        require(
+            totalContributed[msg.sender] > 0,
+            "PartyBid::bid: only contributors can bid"
+        );
+        require(
+            address(this) !=
+                IMarketWrapper(marketWrapper).getCurrentHighestBidder(
+                    auctionId
+                ),
+            "PartyBid::bid: already highest bidder"
+        );
+        require(
+            !IMarketWrapper(marketWrapper).isFinalized(auctionId),
+            "PartyBid::bid: auction already finalized"
+        );
+        // get the minimum next bid for the auction
+        uint256 _bid = IMarketWrapper(marketWrapper).getMinimumBid(auctionId);
+        // ensure there is enough ETH to place the bid including PartyDAO fee
+        require(
+            _bid <= getMaximumBid(),
+            "PartyBid::bid: insufficient funds to bid"
+        );
+        // submit bid to Auction contract using delegatecall
+        (bool success, bytes memory returnData) =
+            marketWrapper.delegatecall(
+                abi.encodeWithSignature("bid(uint256,uint256)", auctionId, _bid)
+            );
+        require(
+            success,
+            string(
+                abi.encodePacked(
+                    "PartyBid::bid: place bid failed: ",
+                    returnData
+                )
+            )
+        );
+        // update highest bid submitted & emit success event
+        highestBid = _bid;
+        emit Bid(_bid);
+    }
+
+    // ======== External: Finalize =========
+
+    /**
+     * @notice Finalize the state of the auction
+     * @dev Emits a Finalized event upon success; callable by anyone
+     */
+    function finalize() external nonReentrant {
+        require(
+            partyStatus == PartyStatus.AUCTION_ACTIVE,
+            "PartyBid::finalize: auction not active"
+        );
+        // finalize auction if it hasn't already been done
+        if (!IMarketWrapper(marketWrapper).isFinalized(auctionId)) {
+            IMarketWrapper(marketWrapper).finalize(auctionId);
```

Comment:
hmmmm in the examples I've seen so far, `finalize` has always been permissionless. it could be more future-proof to make this a delegate call. i think it's a bigger change than i really want to make right now 🤔